### ISSUE TITLE
[Feature]: Support Huggingface TGI 'grammar' param

### ISSUE THREAD

**Issue by krrishdholakia** [BerriAI/litellm#4475]
### The Feature

This is the equivalent of json mode on openai and allows for enforcing the response schema 

https://huggingface.co/docs/text-generation-inference/en/basic_tutorials/using_guidance#grammar-and-constraints-

### Motivation, pitch

better provider support

**Comment by lhoestq** [comment]
"grammar" is also a parameter in llama.cpp, in particular in their [server](https://github.com/ggml-org/llama.cpp/blob/master/examples/server/README.md#:~:text=grammar%3A%20Set%20grammar%20for%20grammar%2Dbased%20sampling.%20Default%3A%20no%20grammar)

**Comment by ghost** [comment]
hey @lhoestq isn't the issue for llama.cpp solved by calling their openai-compatible endpoint?

**Comment by lhoestq** [comment]
yes indeed ! but I like using litellm or tools built on top of litellm better

And tools like https://github.com/lotus-data/lotus are based on litellm but can't benefit from the grammar parameter, e.g. to do reasoning + structured generation using deepseek (I opened https://github.com/lotus-data/lotus/issues/93 a few weeks ago about this)

**Comment by ghost** [comment]
@lhoestq  can't you use this route - https://docs.litellm.ai/docs/providers/openai_compatible? 

route - `openai/`

if llama.cpp's server supports the openai equivalent params, i believe this is solved right?

**Comment by lhoestq** [comment]
Ah the `openai/` route does pass the `grammar` parameter as kwargs to the server ! I didn't know.

```python
resp = litellm.completion(
    model="openai/DeepSeek-R1-Distill-Qwen-1.5B-GGUF", api_key="foo",
    api_base="http://localhost:8080/v1",
    messages=[{"role": "user", "content": "rate this greeting from 1 to 5: 'hello!'"}],
    grammar=grammar,
)
```

Thanks, it works well :)

Anyway I kinda highjacked this issue which is on TGI originally, sorry about that.
I was also confused because TGI uses the `grammar` parameter for json actually, while llama.cpp uses it for formal grammar (GBNF)
